compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -832,7 +832,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 err.subdiagnostic(errors::ExpectedReturnTypeLabel::Unit { span });
                 return true;
             }
-            &hir::FnRetTy::DefaultReturn(span) if expected.is_unit() => {
+            // Don't suggest adding return types for closures with default return.
```

Comment:
There's some redundancy between this check and the `can_suggest` boolean above. Can you find out if we can just reuse `can_suggest` in this position, rather than making the check from within the function?

Reply:
I did some investigations and I found that 

https://github.com/rust-lang/rust/blob/bf662eb80838008acabc307dd64d84935ce3a20d/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs#L940

If I set `can_suggest` to `false` for closures here, then for `tests/ui/closures/add_semicolon_non_block_closure.rs`

https://github.com/rust-lang/rust/blob/bf662eb80838008acabc307dd64d84935ce3a20d/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs#L829-L835

code here will go to the upper branch. I think we still need some way to distinguish `main` and closures or we will get something like this:

```
 --> testfile.rs:8:12
  |
8 |     foo(|| bar())
  |           -^^^^^ expected `()`, found `i32`
  |           |
  |           expected `()` because of default return type
  |
```

I don't expect the last line to be printed.

I've thought of some possible solutions:
- The above result is acceptable?
- Keep my original solution, which indeed adds to complexity
- Add `is_main` to the return values of `get_fn_decl`, which seems (I think) a bit overkill and may impact performance if it's getting called a lot
- Refactor some of the code to see if we can find a better solution